T34576 use p4a loading webview

dylanmccall opened the pull request:
https://phabricator.endlessm.com/T34576

-----

This requires that we merge endlessm/python-for-android#17, first. Note that there is a WIP commit (5ff4729f0ba80e49e875f42f798ed5358ead6d4a) which can be dropped once we have done that.

dylanmccall committed:
Wait for redirects before showing main webview

The onPageFinished callback fires when a page is redirected to another
one. We can detect if a new page load is started before the previous
page is finished, and wait for that new page load to finish.

https://phabricator.endlessm.com/T34576

dylanmccall committed:
Set app key cookie directly

Instead of using the "initialize" URL, we can use the WebKit cookie
manager to immediately set the app key cookie. This reduces the
complexity of starting the application.

https://phabricator.endlessm.com/T34576

dylanmccall committed:
Update python-for-android to v2022.09.04-endless9

https://phabricator.endlessm.com/T34576

dylanmccall committed:
Use a different webview for the loading / welcome screen

https://phabricator.endlessm.com/T34576

manuq commented on src/java/org/learningequality/FullScreen.java:
```diff
@@ -41,37 +44,55 @@ public static FullScreen getInstance() {
     public void configure(final Runnable startWithNetwork, final Runnable startWithUSB, final Runnable loadingReady) {
-        mWebView.setWebContentsDebuggingEnabled(true);
-        mWebView.setWebViewClient(new WebViewClient() {
+        Log.i("Endless Key", "FullScreen configure");
```
Nit: Maybe we can make "Endless Key" a constant:
```
private static final String TAG = "Endless Key";

...

Log.i(TAG, "FullScreen configure");

```

dylanmccall replied:
:+1:

manuq commented on requirements.txt:
```diff
@@ -1,3 +1,3 @@
 cython
 virtualenv
-git+https://github.com/endlessm/python-for-android@v2022.09.04-endless8#egg=python-for-android
+git+https://github.com/endlessm/python-for-android@51f2107a62248feb30f3d21338674fbfb6176184#egg=python-for-android
```
2nd commit is great for testing the p4a changes along with these ones in our CI build! I will approve without merging so you can update this line after tagging in the p4a repo.

dylanmccall replied:
Okay, I changed this to refer to `v2022.09.04-endless9` in the newest push.

manuq commented on src/kolibri_android/main_activity/activity.py:
```diff
@@ -35,8 +37,23 @@ def replace_url_in_webview(url):
+
+@Runnable
+def evaluate_javascript_in_main_webview(js_code):
+    FullScreen.getInstance().mMainWebView.evaluateJavascript(js_code, None)
```
There aren't any users of this one. Let's remove it for now.

manuq commented on src/kolibri_android/main_activity/activity.py:
```diff
@@ -35,8 +37,23 @@ def replace_url_in_webview(url):
+    FullScreen.getInstance().showLoadingPage(url)
+
+
+@Runnable
+def evaluate_javascript_in_loading_webview(js_code):
+    FullScreen.getInstance().mLoadingWebView.evaluateJavascript(js_code, None)
```
:+1:

manuq commented on src/kolibri_android/main_activity/activity.py:
```diff
@@ -145,7 +164,7 @@ def on_activity_save_instance_state(self, activity, out_state_bundle):
 
     def _get_current_kolibri_path(self):
-        current_url = PythonActivity.mWebView.getUrl()
+        current_url = FullScreen.getInstance().getUrl()
```
Nit: Maybe is a good time to rename "FullScreen" to a better name?

dylanmccall replied:
Okay, after some deliberation I renamed it to `KolibriAndroidHelper`. Not a great name, but at least it isn't misleading :)

manuq commented on src/kolibri_android/main_activity/activity.py:
```diff
@@ -154,8 +173,7 @@ def _get_current_kolibri_path(self):
             return None
 
     def run(self):
-        self.replace_url("file:///android_asset/_load.html")
```
:+1:

manuq commented on src/kolibri_android/main_activity/kolibri_bus.py:
```diff
@@ -65,12 +68,6 @@ def register_share_file_interface(share_file):
-        initialize_url = interface.get_initialize_url(next_url=next_url)
-        logging.info(f"Using initialize URL '{initialize_url}'")
-        start_url = urljoin(base_url, initialize_url)
+        next_url = self.application.get_saved_kolibri_path() or ""
+        start_url = urljoin(base_url, next_url)
```
:+1:

dylanmccall committed:
Use a constant for log message tags

https://phabricator.endlessm.com/T34576

dylanmccall committed:
Rename FullScreen to KolibriAndroidHelper

This class has many different functions related to interfacing with the
Android system. We should name it accordingly to avoid confusion.

https://phabricator.endlessm.com/T34576
